### DISCUSSION TITLE
How do you use peripheral Interrupts?

### DISCUSSION THREAD

**Discussion by ProxyPlayerHD** [stm32duino/Arduino_Core_STM32#2107] [Q&A]
I got a project that i want to port from AVR to STM32.
that project mainly works around the MCU being an SPI Slave with interrupt driven communication.
it used the Dx Core which makes peripheral interrupts really simple to use.
you just manually enable the interrupt by writing to the correct control register and then use the `ISR(x)` function with `x` being the name of the vector to use, which depends on the peripheral.

so for my SPI Slave IRQ handler the function was called `ISR(SPI_STC_vect)`

now of course the same exact function won't work in this core as well. but either i'm blind or the wiki doesn't mention an equivalent function... or really ANYTHING about using interrupts.
so how do you do stuff like this with this core?

i can enable the interrupt flag in the SPI control register, that one is easy. but how do i assign a function to run when that specific interrupt hits?

**Comment by ProxyPlayerHD** [comment]
ok i think i got a bit closer to what i want, though it's still not functional.

I Initalize the SPI2 peripheral in the setup function like this:
```
SPI2->DR = 0xFF;			// Preload an 0xFF Byte
	
SPI2->CR2 = 0b0000000010000000;		// Enable the TX Empty Interrupt
SPI2->CR1 = 0b0000000001000000;		// Enable SPI in Slave mode
```

then i (hopefully) set the correct bits in the NVIC like so:
```
NVIC->IP[SPI2_IRQn] = (1U << 4U);				// Set SPI2 Interrupt Priority to 1 (shifted by 4 bits because only the upper 4 bits are implemented)
NVIC->ISER[SPI2_IRQn >> 5U] |= (1U << (SPI2_IRQn & 31U));	// And Enable the SPI2 Interrupt
```

this should trigger an interrupt when the SPI Slave has started to transfer that preloaded 0xFF byte.
i then looked around the repo and found the [startup.s](https://github.com/stm32duino/Arduino_Core_STM32/blob/main/variants/STM32F1xx/F103C8T_F103CB(T-U)/startup_M200_f103xb.S) file that contains the interrupt vectors, and saw that they use weak aliases so you should be able to define your own interrupt functions and they should have their address placed in the vector table as long as they have the same name.

so i did that as well and replaced the old `ISR(SPI_STC_vect)` with `void SPI2_IRQHandler(void)` (with an extern "C" around the whole function)

but still, the interrupt never gets triggered so i don't know which part of this i screwed up as it all seems correct to me. maybe i overlooked something with the NVIC?

any help would be nice :(

**Reply by fpistm** [reply]
Hi @ProxyPlayerHD 
Unfortunately, I have no clue. Didn't check all your register access but seems good.
In a general way, I use HAL or LL API to avoid forget something when using IT.
